[KIP-149] Unified system contract management process (#150)
klaytn/kips: KIPs/kip-149.md

- System contract upgrade: The process of updating an existing logic contract while maintaining a proxy contract. The upgraded logic contract must be compatible with the previous interface.

- System contract replacement: The deployment of a new system contract that is then registered to the Registry using the same name as its predecessor. The new contract effectively deprecates the previous system contract.

### Smart Contracts Overview

The proposed smart contract will be implemented in Solidity and compatible with the Ethereum Virtual Machine (EVM).

The smart contract will have the following features:

1. Registry

    - Register a new system contract with an activation block.

    - Return the state of the system contracts.

2. Proxy

    - Delegate a call to logic contract.

    - Upgrade a logic contract.

#### 1. Registry

The registry should have data for the pre-deployed system contracts as soon as it is deployed. It will be done by state injection, which injects data into the Registry directly using the `state.SetState`. A reference implementation is introduce in [Implementation](#implementation).

#### Interface of Registry

```solidity
pragma solidity ^0.8.0;

> interface Registry {

Reviewer: The specification must stipulate the storage variables because they are crucial to the hardfork consensus.
See https://github.com/klaytn/klaytn/blob/cd09595fe98c768526089d080731e40787b15dcb/contracts/system_contracts/registry/IRegistry.sol.

Laying out storage variables as part of the contract interface is often used technique. For instance, https://github.com/compound-finance/compound-protocol/blob/master/contracts/Governance/GovernorBravoInterfaces.sol#L63 has a separate GovernorBravoDelegateStorageV1 contract, to make sure that implementations retain the exact storage layout.

Author: Since the attached example is for an upgradeable contract (when there's a future possibility to add storage variables), it needs storage layout interface. But I think the Registry doesn't need that much. Do you expect that there can be a case that we need to add more variables for the Registry contract after initial deployment?

Author: But moving the storage variable declaration to the interface for clarification looks necessary.

++ Reflected in https://github.com/klaytn/kips/pull/150/commits/bfd4c7eecae354500d4bd62b2bcd5436e7f0c488

Reviewer: @hyeonLewis thanks. a few more fixes: (1) `interface` cannot contain variables. make it `abstract contract`. (2) variable 'records' and automatic getter function 'records' collides. remove the function. (3) can we rename 'contractNames' to 'names'? because we have 'getAllNames' getter and argument names 'string name'.

Author: Oh, my bad. Fixed at https://github.com/klaytn/kips/pull/150/commits/b80e5e7f1d43b969219f4e267e2483858c52c268

![](../assets/kip-149/ReplacementProcess.png)

### Core Logic Overview

After a target block number, a Klaytn node should read all the active addresses of system contracts through the Registry. A Klaytn node deploys the Registry at the configured block number at the reserved address by bytecode injection.

#### Chain Configuration

In the Chain Config, the following field is introduced. All node operators in a network must update `genesis.json` configuration with the same value. The configuration values for Baobab and Cypress networks are hard-coded on the client source code.

- `kip149CompatibleBlock`: the target block number that a Klaytn node deploys Registry.

#### Execution

The Registry deployment is executed at the `Finalize` function, which means the end of the block generation. It reads the reserved address and runtime bytecode and deploys the Registry by the bytecode injection. Also, it injects the state for pre-deployed system contracts here.

```go
// Deploy the registry contract if the block is the kip-149 fork block

> if chain.Config().IsKIP149ForkBlock(header.Number) {

Reviewer: It is installed BEFORE the fork block. See https://github.com/klaytn/klaytn/pull/1987.

Author: Reflected in https://github.com/klaytn/kips/pull/150/commits/bfd4c7eecae354500d4bd62b2bcd5436e7f0c488

author: Lewis (@hyeonLewis), Ian (@ian0371), Ollie (@blukat29), Lake (@hyunsooda), and Aidan (@aidan-kwon)
discussions-to: https://github.com/klaytn/kips/issues/149
status: Draft
type: Standards Track
category: Core
created: 2023-09-20
---

## Simple Summary

A unified deployment and management process for system contracts.

## Abstract

This standard defines a unified deployment and management process for system contracts. To effectively manage system contracts, it also introduces a Registry contract that manages all system contracts.

## Motivation

Currently, system contracts are deployed and managed without any defined standards. For example, the `AddressBook` contract was deployed by a bytecode injection at the genesis block with a reserved address, while EOA deployed `TreasuryRebalance`, and its address is set in the chain config. As more system contracts will be deployed in the future, it’s essential to have a standard way to deploy and manage system contracts.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> This standard defines the separation of data and logic contracts in system contracts. This method, often called the proxy pattern, allows the change of the logic contract while keeping the data, which can greatly reduce the cost of contract updates. Upgrading a logic contract will not affect the Registry since the Registry only holds the address of the proxy(data) contract. Delegating ownership of a proxy contract to a governance contract can solve the centralized issue and potential private key loss.

Reviewer 2: Does this imply that all system contracts henceforth must be deployed with a proxy? Whether it's true or not, I think it should be clearly stated.

Author: Reflected in https://github.com/klaytn/kips/pull/150/commits/b872f0e9bb5457a0f7a56de0d2db274edac81066

Reviewer 2: I think `will be` should be either `must be` or `should be`

Author: `should be` looks good to me. Amended at https://github.com/klaytn/kips/pull/150/commits/b8eeffe54983cdb11c0d62bf04fc17ae2ba391a8

> This standard defines the separation of data and logic contracts in system contracts. This method, often called the proxy pattern, allows the change of the logic contract while keeping the data, which can greatly reduce the cost of contract updates. Upgrading a logic contract will not affect the Registry since the Registry only holds the address of the proxy(data) contract. Delegating ownership of a proxy contract to a governance contract can solve the centralized issue and potential private key loss.

Reviewer 2: Also, could you add a section in Rationale that explains the implication of Governance being the owner of Registry? It should contain (1) delegating the right to upgrade/replace system contracts to GCs and validators, (2) so we need a consent from GCs and validators, (3) the consent from GCs is achieved via voting, (4) the consent from validators is achieved by their waiving the right to veto (i.e., hardfork), (5) private key issue mentioned here

Author: Reflected in ~https://github.com/klaytn/kips/pull/150/commits/b872f0e9bb5457a0f7a56de0d2db274edac81066~ https://github.com/klaytn/kips/pull/150/commits/b8eeffe54983cdb11c0d62bf04fc17ae2ba391a8

- Active: The current block number exceeds its activation, and no active successor system contract exists.

- Deprecated: There’s an active successor system contract.

![](../assets/kip-149/LifeCycle.png)

#### Upgrade System Contracts

When upgrading system contracts, its logic contract will be changed by governance proposal. The Registry will not be updated since it only manages the address of the proxy contract.

![](../assets/kip-149/UpgradeProcess.png)

#### Replace System Contracts

If current system contract updates can’t be done by upgrading the logic contract, it must be replaced with the newly deployed system contract. The predecessor doesn’t need to be explicitly deprecated since a new system contract will implicitly replace and deprecate it. The replacement process is the same as the initial registration for the system contract except for having a predecessor.

![](../assets/kip-149/ReplacementProcess.png)

### Core Logic Overview

After a target block number, a Klaytn node should read all the active addresses of system contracts through the Registry. A Klaytn node deploys the Registry at the configured block number at the reserved address by bytecode injection.

> #### Chain Configuration

Reviewer 2: Could you add the address of Registry 0x401 in this section?

Author: Reflected in ~https://github.com/klaytn/kips/pull/150/commits/b872f0e9bb5457a0f7a56de0d2db274edac81066~ https://github.com/klaytn/kips/pull/150/commits/b8eeffe54983cdb11c0d62bf04fc17ae2ba391a8

## Motivation

Currently, system contracts are deployed and managed without any defined standards. For example, the `AddressBook` contract was deployed by a bytecode injection at the genesis block with a reserved address, while EOA deployed `TreasuryRebalance`, and its address is set in the chain config. As more system contracts will be deployed in the future, it’s essential to have a standard way to deploy and manage system contracts.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

This standard defines the separation of data and logic contracts in system contracts. This method, often called the proxy pattern, allows the change of the logic contract while keeping the data, which can greatly reduce the cost of contract updates. Upgrading a logic contract will not affect the Registry since the Registry only holds the address of the proxy(data) contract. Delegating ownership of a proxy contract to a governance contract can solve the centralized issue and potential private key loss.

### Definitions

> - System contract: A contract that affects protocol. The currently deployed system contracts are as follows: **AddressBook**, **Governance**([KIP-81](https://github.com/klaytn/kips/blob/main/KIPs/kip-81.md)), **StakingTracker**, **TreasuryRebalance**([KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md))

Reviewer 3: ```suggestion
- System contract: A contract that affects protocol. The currently deployed system contracts are as follows: **AddressBook**, **Governance**([KIP-81](https://github.com/klaytn/kips/blob/main/KIPs/kip-81.md)), **StakingTracker**, **Voting**, **TreasuryRebalance**([KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md))
```

Author: My mistake. Changing `governance` to `voting`. Applied in https://github.com/klaytn/kips/pull/150/commits/fa942df119da53daef6c6b2b24a2e80aa47aded4

> - System contract upgrade: Upgrade a logic contract while maintaining a proxy contract. The upgraded logic contract must maintain the previous interface.

Reviewer 4: - System contract upgrade: The process of updating an existing logic contract while maintaining a proxy contract. The upgraded logic contract must maintain the previous interface.

Author: Applied in https://github.com/klaytn/kips/pull/150/commits/fa942df119da53daef6c6b2b24a2e80aa47aded4

> - System contract upgrade: Upgrade a logic contract while maintaining a proxy contract. The upgraded logic contract must maintain the previous interface.

Reviewer 3: >  The upgraded logic contract must maintain the previous interface.

The sentence can be readable if the word `maintain` is replaced with `compatible`. What does 'maintain' mean? It may imply the lifetime of a legacy contract is still valid, but I guess it would not. Could you clarify this?

Author: Applied in https://github.com/klaytn/kips/pull/150/commits/fa942df119da53daef6c6b2b24a2e80aa47aded4 (With [Scott's](https://github.com/klaytn/kips/pull/150#discussion_r1333696519) review)

> - System contract replacement: Deploy a new system contract and register it to the Registry with the same name. It will deprecate predecessor system contract.

Reviewer 4: - System contract replacement: The deployment of a new system contract that is then registered to the Registry using the same name as its predecessor. The new contract effectively deprecates the previous system contract.

Author: Applied in https://github.com/klaytn/kips/pull/150/commits/fa942df119da53daef6c6b2b24a2e80aa47aded4

### Smart Contracts Overview

The proposed smart contract will be implemented in Solidity and compatible with the Ethereum Virtual Machine (EVM).

The smart contract will have the following features:

1. Registry

    - Register a new system contract with an activation block.

>     - Return the information of system contracts.

Reviewer 3: `state` seems a better fit than `information`.

Author: Applied in https://github.com/klaytn/kips/pull/150/commits/fa942df119da53daef6c6b2b24a2e80aa47aded4

#### Chain Configuration

In the Chain Config, the following field is introduced. All node operators in a network must update `genesis.json` configuration with the same value. The configuration values for Baobab and Cypress networks are hard-coded on the client source code.

- `kip149CompatibleBlock`: the target block number that a Klaytn node deploys Registry.

#### Execution

The Registry deployment is executed at the `Finalize` function, which means the end of the block processing process. It reads the reserved address and runtime bytecode and deploys the Registry by the bytecode injection.

```go
// Deploy the registry contract if the block is the kip-149 fork block
if chain.Config().IsKIP149ForkBlock(header.Number) {
	// Set the registry contract code to the stateDB
	err := registry.InstallRegistry(state)
	if err != nil {
		logger.Error("failed to set the registry contract code", "err", err)
	} else {
		logger.Info("successfully set the registry contract code", "block", header.Number.Uint64())
	}
}
```

#### Resolver

A Klaytn node will have a resolver to read the active addresses of system contracts from the Registry.

## Rationale

> ### Bytecode Injection For Registry Deployment

Reviewer 4: ### Bytecode Injection for Registry Deployment

Author: Applied in https://github.com/klaytn/kips/pull/150/commits/fa942df119da53daef6c6b2b24a2e80aa47aded4

In the future, all system contracts will be registered in the Registry, and a Klaytn node will read the active addresses of system contracts from the Registry. Not only for a Klaytn node but also for other ecosystem participants who will use the registry to read the system contracts they need, meaning the registry should be registered at an easily accessible reserved address.

> ### Separate Data And Logic Contract

Reviewer 4: ### Separate Data and Logic Contract

Author: Applied in https://github.com/klaytn/kips/pull/150/commits/fa942df119da53daef6c6b2b24a2e80aa47aded4

This proxy pattern simplifies the process of system contract update because the existing data can be used even if the logic contract is changed. The main issue of the proxy pattern is the centralization and potential private key loss. But delegating ownership to upgrade a logic contract to Governance can solve those problems.

## Backward Compatibility

### Deployed System Contracts

As explained in [Registry](#1-registry) section, it consists of base and network-specific registries. The below code is an example of `RegistryCypress`.

```solidity
// Pre-deployed system contracts
address public constant ADDRESS_BOOK = 0x0000000000000000000000000000000000000400;
address public constant VOTING = 0xcA4Ef926634A530f12e55A0aEE87F195A7B22Aa3;
address public constant TREASURY_REBALANCE = 0xD5ad6D61Dd87EdabE2332607C328f5cc96aeCB95;
address public constant STAKING_TRACKER = 0x9b8688d616D3D5180d29520c6a0E28582E82BF4d;

bytes32 public constant ADDRESS_BOOK_KEY = keccak256(abi.encodePacked("AddressBook"));
bytes32 public constant VOTING_KEY = keccak256(abi.encodePacked("Voting"));
bytes32 public constant TREASURY_REBALANCE_KEY = keccak256(abi.encodePacked("TreasuryRebalance"));

> bytes32 public constant STAKING_TRACKER_KEY = keccak256(abi.encodePacked("StakingTracker"));

Reviewer 3: Is the `GovParam` contract intentionally excluded from management within the `Registry` contract?

Author: I'm not sure whether `GovParam` is deployed or not. (AFAIK, not deployed and used) If deployed, then we need to include it in the wrapper. But if not, we can register `GovParam` in the `register` function when needed.

Reviewer 3: This comment is not valid at this point as the usage was revised.

The `GovParam` contract was deployed on both Cypress and Baobab netowrk. You can query the address through `klay.getParams()`.

---
kip: 149
title: Unified System Contract Management Process
author: Lewis (@hyeonLewis), Ian (@ian0371), Ollie (@blukat29), Lake (@hyunsooda), and Aidan (@aidan-kwon)
discussions-to: https://github.com/klaytn/kips/issues/149
status: Draft
type: Standards Track
category: Core
created: 2023-09-20
---

## Simple Summary

A unified deployment and management process for system contracts.

## Abstract

This standard defines a unified deployment and management process for system contracts. To effectively manage system contracts, it also introduces a Registry contract that manages all system contracts.

## Motivation

> Currently, system contracts are deployed and managed without any defined standards. For example, the `AddressBook` contract was deployed by a bytecode injection at the genesis block with a reserved address. In contrast, EOA deployed `TreasuryRebalance`, and its address is set in the chain config. As more system contracts will be deployed in the future, it’s essential to have a standard way to deploy and manage system contracts.

Reviewer 5: ```suggestion
Currently, system contracts are deployed and managed without any defined standards. For example, the `AddressBook` contract was deployed by a bytecode injection at the genesis block with a reserved address. In contrast, an EOA deployed `TreasuryRebalance`, and its address is set in the chain config. As more system contracts will be deployed in the future, it’s essential to have a standard way to deploy and manage system contracts.
```

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/93c6dd56fc49084969b179a184f473c4b7683baf

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> This standard defines the separation of data and logic contracts in system contracts, and every system contract should be developed and deployed under this standard. This method, often called the proxy pattern, allows the change of the logic contract while keeping the data, which can greatly reduce the cost of contract updates. Upgrading a logic contract will not affect the Registry since the Registry only holds the address of the proxy(data) contract. Delegating ownership of a proxy contract to a governance contract can solve the centralized and potential private key loss issues.

Reviewer 5: Can we define what are "system contracts"?

Reviewer 5: oh, it is defined later. It would be good to have a link to its definition.

Author: Linked to the first occurrence of system contracts in the `Motivation` table since it looks more natural I think. Applied at https://github.com/klaytn/kips/pull/150/commits/f64b7b9f3323d6fb7c66e7970522247bd6f58ed8

Between the popular proxy patterns `Transparent` and `UUPS`, all system contracts must be developed and validated based on the [`UUPS`](https://eips.ethereum.org/EIPS/eip-1822) proxy pattern.

### Definitions

> - System contract: A contract that is read by a Klaytn node or directly affect the protocol. The currently deployed system contracts are as follows: **AddressBook**, **GovParam**, **Voting**([KIP-81](https://github.com/klaytn/kips/blob/main/KIPs/kip-81.md)), **TreasuryRebalance**([KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md))

Reviewer 5: ```suggestion
- System contract: A contract that is read by the Klaytn core protocol or directly affect the protocol. The currently deployed system contracts are as follows: **AddressBook**, **GovParam**, **Voting**([KIP-81](https://github.com/klaytn/kips/blob/main/KIPs/kip-81.md)), **TreasuryRebalance**([KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md))
```

all contract can be read by a klaytn node. 😂

Reviewer 5: Can we have also a link to the contracts "AddressBook" and "GovParam"? If we don't have any KIP for that, we can also make a link to the source code directly.

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

> - System contract upgrade: The process of updating an existing logic contract while maintaining a proxy contract. The upgraded logic contract must be compatible with the previous interface.

Reviewer 5: ```suggestion
- System contract upgrade: The process of updating an existing logic contract while maintaining a proxy contract. The upgraded logic contract must be compatible with the previous interface and its storage layout.
```

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

> - System contract replacement: The deployment of a new system contract that is then registered to the Registry using the same name as its predecessor. The new contract effectively deprecates the previous system contract.

Reviewer 5: In this case, do we need to deploy both logic and proxy contract?

Author: Yes, `system contract replacement` exists to handle the case that the logic contract cannot be upgraded anymore (breaking changes, different storage layout...)

### Smart Contracts Overview

The proposed smart contract will be implemented in Solidity and compatible with the Ethereum Virtual Machine (EVM).

The smart contract will have the following features:

1. Registry

>     - Register a new system contract with an activation block.

Reviewer 5: Do we need to specify the replacement feature? or is it enough to cover replacement with this registration feature?

Author: I thought it would be enough since the `Replacement` is more like a "process", not the "job" of a Registry contract.

    /// @dev Array of system contract names
    string[] public names;

    /// @dev Owner of contract
    address internal _owner;

    /* ========== TYPES ========== */
    /// @dev Struct of system contracts
    struct Record {
        address addr;
        uint256 activation;
    }

    /* ========== EVENTS ========== */
    /// @dev Emitted when the contract owner is updated by `transferOwnership`.
    event OwnershipTransferred(address indexed previousOwner, address indexed newOwner);
    
    /// @dev Emitted when a new system contract is registered.
    event Registered(string name, address indexed addr, uint256 indexed activation);

    /* ========== MUTATORS ========== */
    /// @dev Registers a new system contract.
    function register(string memory name, address addr, uint256 activation) external virtual;
    
    /// @dev Transfers ownership to newOwner.
    function transferOwnership(address newOwner) external virtual;

    /* ========== GETTERS ========== */
    /// @dev Returns an address for active system contracts registered as name if exists.
    ///  It returns a zero address when there's no active system contract with name.

>     function getActiveAddr(string memory name) external virtual returns (address);

Reviewer 5: Is there any reason that this function does not have the `view` attribute?

Author: Oh, it's a typo. Thanks. Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

  
    /// @dev Returns all system contracts registered as name.
    function getAllRecords(string memory name) external virtual view returns (Record[] memory);
    
    /// @dev Returns all names of registered system contracts.
    function getAllNames() external virtual view returns (string[] memory);
    
    /// @dev Returns owner of contract.
    function owner() external virtual view returns (address);
}
```

#### Methods

```solidity
function register(string memory name, address addr, uint256 activation)
```

Registers a new system contract. It will be activated at `activation`. It overwrites the predecessor if a predecessor system contract exists and is not yet active. Passing `addr == address(0)` is an implicit deprecation for the `name`, meaning the `name` will no longer be used.

The function validates the following requirements:

> - The function caller MUST be a governance contract.

Reviewer 5: How can we determine the caller is the governance contract? I don't see any declaration of a variable having the governance contract.

Author: Yes, I think `owner` looks better. Is it enough to change `governance contract` to `owner address`?

Author: Yes, I think `owner` looks more fit (it'll be a governance contract in the future). I'll update it to `owner address`.

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/779d283e714c2c3c7a84d383e84b9574976e4402

> - The function MUST revert if a `name` is an empty string.

Reviewer 5: How can we detect typo? Our intention was 'AddressBook', but we put it 'addressBook'.

Please ignore this comment if you think it makes more confusion, but my suggestion is:
1. we register specific names first.
2. Then we can only register the predefined names only.

Author: I agree with your comment, but a typo is difficult to occur and easy to spot and recover. So, there's not much to be gained by implementing it.

> 
> - The function MUST revert if `activation < block.number`.
> 

Reviewer 5: ```suggestion

- The function MUST revert if `addr` is in the range of precompiled contract address range.

- The function MUST revert if `activation < block.number`.
```

Reviewer 5: Is it okay to activate immediately? It would be also good to have some buffer to activation.

Also there is the case we need to force to activate immediately in an emergency situation. It is very hard to setup exact block number. So I think it would be also good to prepare two functions.
1. normal function with at least 100 block prior to activate
2. emergency function with immediate activation.

Author: 1. Why do we need to revert if it's in the range of precompiled contract range? Since we don't know the future usage of precompiled contracts, I think it's better not to prohibit them in KIP explicitly.

2. Good point. We also considered many cases in various conditions but determined the current design since:
    a. System contract registration will be done with much consideration and governance(KIP-81) discussion in the future, and activation will be given with enough buffer if it goes with a hard fork or needs specific block timing.
    b. Emergency activation/deprecation or any methods may be considered a backdoor.

If you think of any needed buffer or emergency functions, it's better to discuss them with other core team members. cc @blukat29 @ian0371 .

Reviewer 2: 1. I think the buffer from governance (Voting) is enough; if we introduce another buffer in this contract, it's hard to predict and set the activation number prior to governance.

2. Speaking of emergency function, as long as the caller is Voting, I think it wouldn't hurt to have a function like GovParam's `setParamIn`, which can be used to set the activation block relatively based on the current block.

Reviewer 5: I'm not exactly sure whether this is needed or not. Based on the communication with @ian0371 and @hyeonLewis, it seems like it is not necessary to include. If any other opinion, please let me know.

The function emits a `Registered` event.

```solidity
function getActiveAddr(string memory name) view returns (address)
```

Returns the address of the active system contract with the `name`. It returns a zero address if there’s no registered system contract with the `name` or the `name` has been deprecated by registering a zero address.

#### 2. Proxy

The implementation of the proxy contract will come from [OZ's UUPS implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.3/contracts/proxy/ERC1967/ERC1967Proxy.sol), which follows [EIP-1967](https://eips.ethereum.org/EIPS/eip-1967) and [EIP-1822](https://eips.ethereum.org/EIPS/eip-1822). 

> ### System Contracts Life Cycle

Reviewer 5: Is it for a specific system contract or for a system contract name?

Author: It's for both system contract and name since every system contract after KIP-149 will follow this and the Registry manages the system contracts with their name. But I think it's closer to the `name` in this KIP since it's for system contracts management in the Registry contract.

The Registry contract manages system contracts based on the current block number. Its state will be managed implicitly, which means there’s no explicit state variable(e.g., enum State). It has three implicit states and cannot be reversed to the previous state:

- Registered: It has been registered but has not been activated yet.

- Active: The current block number exceeds its activation, and no active successor system contract exists.

> - Deprecated: There’s an active successor system contract.

Reviewer 5: What about registering '0'?

Author: Does it mean to register `activation = 0`? It might cause confusion since activating from `block=0` makes intuitive sense(and `KIP113` will be `activation = 0`). And we need additional logic to filter out `activation == 0` to search in a view function.

Anyway, I need to add a condition with zero address(deprecated name) for `Active`, and `Deprecated`. Applied at https://github.com/klaytn/kips/pull/150/commits/821b02ac3ff714726d04e949634da90d8524cffb

![](../assets/kip-149/LifeCycle.png)

#### Upgrade System Contracts

> When upgrading system contracts, its logic contract will be changed by governance proposal. The Registry will not be updated since it only manages the address of the proxy contract.

Reviewer 5: ```suggestion
When upgrading a system contract, its logic contract will be changed by a governance proposal. The Registry will not be updated since it only manages the address of the proxy contract.
```

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

![](../assets/kip-149/UpgradeProcess.png)

#### Replace System Contracts

If current system contract updates can’t be done by upgrading the logic contract, it must be replaced with the newly deployed system contract. The predecessor doesn’t need to be explicitly deprecated since a new system contract will implicitly replace and deprecate it. The replacement process is the same as the initial registration for the system contract except for having a predecessor.

![](../assets/kip-149/ReplacementProcess.png)

### Core Logic Overview

> After a target block number, a Klaytn node should read all the active addresses of system contracts through the Registry. A Klaytn node deploys the Registry at the configured block number at the reserved address by bytecode injection.

Reviewer 5: How can the Klaytn node deploy the Registry? Do we need to do hard-fork to deploy Registry? It would be good to elaborate more on how to deploy/replace the Registry contract.

Author: The deployment implementation is written [here](https://github.com/hyeonLewis/kips/blob/kip149/KIPs/kip-149.md?plain=1#L213). It needs hardfork as it says.

I'll add an explanation about the `replace` case. It will be done by bytecode injection, the same as its deployment process. Applied at https://github.com/klaytn/kips/pull/150/commits/c99177fbac7634b2ad21e9aba01e3d6a5bb46c58

#### Chain Configuration

> In the Chain Config, the following field is introduced. All node operators in a network must update `genesis.json` configuration with the same value. The configuration values for Baobab and Cypress networks are hard-coded on the client source code.

Reviewer 5: ```suggestion
In the Chain Config, the following fields are newly introduced. All node operators in a network must update `genesis.json` configuration with the same value. The configuration values for Baobab and Cypress networks are hard-coded on the client source code.
```

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

> - `RegistryAddress`: the reserved address for the Registry contract, which is `0x0000000000000000000000000000000000000401`.

Reviewer 5: How can we define the bytecode of the Registry contract? Isn't it necessary to add in the chain config?

Author: It will be generated by `abigen` and read from [here](https://github.com/klaytn/klaytn/blob/dev/contracts/system_contracts/all.go), the same way as `AddressBook` or `GovParam`.

- `Kip149CompatibleBlock`: the target block number that the Registry will be deployed.

> - `RegistryConfig`: the initial data config for the Registry. It is used when injecting the initial state after deploying the Registry.

Reviewer 5: ```suggestion
- `RegistryInit`: the initial data config for the Registry. The type of this field is defined as `RegistryConfig` as shown below. It is used when injecting the initial state after deploying the Registry contract.
```

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

    // Note that the registry only records the system contracts based on the KIP-149, which is currently only KIP-113
    // The activation block deployed by state injection will be 0
    RegistryInit: &RegistryConfig{
		Records: map[string]common.Address{
			"KIP113": Kip113Address,
		},
		Owner: OwnerAddress,
    },
}
```

#### Execution

The Registry deployment is executed at the `engine.Finalize` function, which means the end of the block generation. It reads the reserved address and runtime bytecode and deploys the Registry by the bytecode injection. Also, it injects the initial state provided by `RegistryConfig` for the Registry here. When `kip149CompatibleBlock == 0`, the Registry will be allocated at the genesis block.

```go
if chain.Config().IsKIP149ForkBlock(header.Number) {
	// Inject the bytecode and states for the Registry
	err := registry.InstallRegistry(state, chain.Config().RegistryInit)
	if err != nil {
		logger.Error("failed to set the registry contract code", "err", err)
	} else {
		logger.Info("successfully set the registry contract code", "block", header.Number.Uint64())
	}
}
```

#### Resolver

> A Klaytn node will have a resolver to read the active addresses of system contracts from the Registry. But the system contracts deployed by state injection (e.g., `KIP113`) will be directly read from `chain.Config().RegistryInit` at the `Kip149CompatibleBlock` since the Registry will be deployed in the `engine.Finalize` function, which is the last part of the block generation. In other words, the resolver will be used starting in the `Kip149CompatibleBlock + 1`.

Reviewer 5: hmm what about deploying the registry contract at the block number of `Kip149CompatibleBlock - 1`?

Author: It was initially planned to deploy at the parent block of `Kip149CompatibleBlock`, but updated to deploy it at fork block. (Check this [PR](https://github.com/klaytn/klaytn/pull/2059) would be helpful Note that the `RandaoForkBlock` works as `Kip149ForkBlock `). The main reasons were:
1. It's more straightforward to do it all in one block.
2. Assume that the Kip149 hard fork makes an error and stops the network. Since we already deployed `Registry` at the parent block, we must roll back to two blocks behind and deactivate the hard fork. But if we put it together, we only need to deactivate hard fork (make it nil or delay it).

```go

> // Note that it will be used starting in the next of kip-149 fork block

Reviewer 5: ```suggestion
// Note that it will be activated in the next of KIP-149 fork block
```

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

func ReadRegistryActiveAddr(backend bind.ContractCaller, name string, num *big.Int) (common.Address, error) {

> 	code, err := backend.CodeAt(context.Background(), RegistryAddr, num)

Reviewer 5: I cannot see any code that this function will be used after Kip149CompatibleBlock.

And the function name is not clear. Maybe `ReadActiveAddressFromRegistry`?

Author: I thought the [explanation](https://github.com/hyeonLewis/kips/blob/kip149/KIPs/kip-149.md?plain=1#L224) is enough. Would it be better to show the exact code to read `ReadRegistryActiveAddr` from the Klaytn client?

I'll update the function name to `ReadActiveAddressFromRegistry` in both KIP and Klaytn client. Applied at https://github.com/klaytn/kips/pull/150/commits/355cce46224c5b226fd0442fa3ab8ac2949961d4

Author: Also PR for Klaytn core: https://github.com/klaytn/klaytn/pull/2061

	if err != nil {
		return common.Address{}, err
	}
	if code == nil {
		return common.Address{}, ErrRegistryNotInstalled
	}

	caller, err := contracts.NewRegistryCaller(RegistryAddr, backend)
	if err != nil {
		return common.Address{}, err
	}

	opts := &bind.CallOpts{BlockNumber: num}
	return caller.GetActiveAddr(opts, name)
}
```

## Rationale

### Bytecode Injection for Registry Deployment

In the future, all system contracts will be registered in the Registry, and a Klaytn node will read the active addresses of system contracts from the Registry. Not only for a Klaytn node but also for other ecosystem participants who will use the registry to read the system contracts they need, meaning the registry should be registered at an easily accessible reserved address.

### Delegating Ownership of Registry and System Contracts to Governance

> The Registry holds all system contracts for Klaytn, affecting the protocol directly. This means its registration and deprecation process must be very careful and not centralized. The same goes for the system contract. As mentioned in [specifictaion](#specification), making the EOA the owner of system contracts can cause centralization and potential private key loss issues. By delegating ownership of the Registry and system contracts to Governance, all registry and system contract changes will only be applied after full discussion and consensus of the GCs and Validators. Validators have the right to accept or reject the proposal depending on whether or not the hard fork proceeds.

Reviewer 5: ```suggestion
The Registry holds all system contracts for Klaytn, affecting the protocol directly. This means its registration and deprecation process must be very careful and not centralized. The same goes for the system contract. As mentioned in [specification](#specification), making the EOA the owner of system contracts can cause centralization and potential private key loss issues. By delegating ownership of the Registry and system contracts to Governance, all registry and system contract changes will only be applied after full discussion and consensus of the GCs and Validators. Validators have the right to accept or reject the proposal depending on whether or not the hard fork proceeds.
```

Author: Applied at https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a

### Bytecode Injection for Registry Deployment

In the future, all system contracts will be registered in the Registry, and a Klaytn node will read the active addresses of system contracts from the Registry. Not only for a Klaytn node but also for other ecosystem participants who will use the registry to read the system contracts they need, meaning the registry should be registered at an easily accessible reserved address.

### Delegating Ownership of Registry and System Contracts to Governance

The Registry holds all system contracts for Klaytn, affecting the protocol directly. This means its registration and deprecation process must be very careful and not centralized. The same goes for the system contract. As mentioned in [specifictaion](#specification), making the EOA the owner of system contracts can cause centralization and potential private key loss issues. By delegating ownership of the Registry and system contracts to Governance, all registry and system contract changes will only be applied after full discussion and consensus of the GCs and Validators. Validators have the right to accept or reject the proposal depending on whether or not the hard fork proceeds.

> ### State Injection for Pre-deployed System Contracts

Reviewer 6: will be updated

![](../assets/kip-149/LifeCycle.png)

1. System Contract Deployed: It has been deployed. It contains connecting proxy and logic contracts and transferring ownership of proxy contract to Governance.

2. Data Setup: It has all the data to be activated.

3. Governance: Its registration has been proposed to Governance and is waiting to be registered if it passes.

4. Registered: It has been registered in the registry but has not been activated yet.

5. Active: It has been activated in the registry and is currently used.

6. Deprecated: It has been deprecated in the registry and is no longer being managed or used.

#### Upgrade System Contracts

When upgrading system contracts, its logic contract will be changed by governance proposal. The Registry will not be updated since it only manages the address of the proxy contract.

![](../assets/kip-149/UpgradeProcess.png)

#### Replace System Contracts

If system contract updates can’t be done by upgrading the logic contract, it must be replaced with the newly deployed contract. It needs two transactions from the governance: 1. Deprecate the predecessor system contract first and 2. Register the new system contract.

> The `successor's activation and the predecessor's deprecation` might differ, but the successor's `activation` block must be higher than the `deprecation` block of the predecessor.

Reviewer 3: I popped up one question suddenly. 
Once replacement is done i.e.,
     - [legacy] activate -> deprecate (zero address)
     - [new] registered -> activate

If the activate block number is not the number of deprecate number + 1, there's a period that does not have an activated system contract. Is there a need for a complement or clarification for this?

Author: It was a legacy part so you can ignore this. Now the Registry doesn't hold `deprecation`. A predecessor system contract can only be deprecated by a new active successor. (It's almost the same as `GovParam`)

Reviewer 3: I pointed out the operation misuse. For instance,

- [legacy contract]  activate (500) -> deprecate (zero address)
- [new contract]      registered -> activate (600)

At block number 550, the activation address is the legacy contract address. However, if the register function is called with the legacy contract name, zero address, and activation 550, the range from 550 to 559 returns a zero address, indicating no active contract for this range.

This operation isn't a bug, but it could lead to unintended consequences due to API misuse. Although this scenario is expected to be rare, I've asked the need for clarification or enhancement to address this potential issue.

Author: Could you give me some examples of `unintended consequences`? I'm trying to understand what kinds of clarification or enhancement are needed other than only returning the result itself (return zero address)

Reviewer 3: I was just concerned about the operation misuse example, depicted above. I believe that governance would not make this case. Let's close this.
